### PR TITLE
[SPARK-46474][INFRA] Upgrade upload-artifact action to v4

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to upgrade upload-artifact action from v3 to v4.
The pr is also an improvement after the reversal of https://github.com/apache/spark/pull/44438.

### Why are the changes needed?
- v4.0.0 release notes: https://github.com/actions/upload-artifact/releases/tag/v4.0.0
They have numerous performance and behavioral improvements.

- v3 VS v4: https://github.com/actions/upload-artifact/compare/v3...v4.0.0

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Pass GA.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -468,15 +468,15 @@ jobs:
         name: PySpark
     - name: Upload test results to report
       if: always()
-      uses: actions/upload-artifact@v3
+      uses: actions/upload-artifact@v4
       with:
-        name: test-results-${{ matrix.modules }}--${{ matrix.java }}-${{ inputs.hadoop }}-hive2.3
+        name: test-results-${{ matrix.modules }}--${{ matrix.java }}-${{ inputs.hadoop }}-hive2.3-${{ env.PYTHON_TO_TEST }}
```
Comment (panbingkun) [inline .github/workflows/build_and_test.yml:473]: This is to fix `the break changes feature` of the new version v4:
https://github.com/actions/upload-artifact?tab=readme-ov-file#breaking-changes
image

File: .github/workflows/build_and_test.yml [hunk 2]
```diff
@@ -468,15 +468,15 @@ jobs:
...
         path: "**/target/test-reports/*.xml"
     - name: Upload unit tests log files
       if: ${{ !success() }}
-      uses: actions/upload-artifact@v3
+      uses: actions/upload-artifact@v4
       with:
-        name: unit-tests-log-${{ matrix.modules }}--${{ matrix.java }}-${{ inputs.hadoop }}-hive2.3
+        name: unit-tests-log-${{ matrix.modules }}--${{ matrix.java }}-${{ inputs.hadoop }}-hive2.3-${{ env.PYTHON_TO_TEST }}
```
Comment (panbingkun) [inline .github/workflows/build_and_test.yml:479]: ditto

**Comment by panbingkun** [comment]
Why does this PR https://github.com/apache/spark/pull/44438 cause upload failure in `build_python` testing?
1.

2.

**Review by dongjoon-hyun — APPROVED** [review]
Thank you, @panbingkun .

cc @HyukjinKwon

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by dongjoon-hyun** [comment]
Oh, it seems to fail again.
- https://github.com/apache/spark/actions/runs/7490604055/job/20389931713
> Error: Failed to CreateArtifact: Received non-retryable error: Failed request: (409) Conflict: an artifact with this name already exists on the workflow run

**Comment by dongjoon-hyun** [comment]
Unfortunately, let me revert this again.

**Comment by dongjoon-hyun** [comment]
This is reverted via https://github.com/apache/spark/commit/c18c87dc70d4dfbb40335fe655f79da21835ff80

**Comment by HyukjinKwon** [comment]
Thanks for reverting it 👍

**Comment by panbingkun** [comment]
I also noticed that, it's strange, I will investigate again.

**Comment by panbingkun** [comment]
image
It seems that the ` PYTHON_TO_TEST` environment variable is not effective, 
`Python 3.9 still` runs in various workflows, which doesn't seem to meet our expectations.

**Comment by panbingkun** [comment]
I have found another issue, we need to use `$PYTHON_TO_TEST` in every step of using `env: ${{ fromJSON(inputs.envs) }}`, otherwise the enumeration will not be as expected. 
Let's take a successful `build_ python` Using as an example, https://github.com/apache/spark/actions/runs/7476792796/job/20348090667

1.python 3.10
https://github.com/apache/spark/actions/runs/7476792796/job/20348079659
image

2.python 3.11
https://github.com/apache/spark/actions/runs/7476792796/job/20348091081
image


cc @dongjoon-hyun @HyukjinKwon 
I will first propose a separate PR to fix this issue.

**Pull request closed without merging**